### PR TITLE
add equality operator to compare meshes and forcings

### PR DESCRIPTION
**Pull request by ghost**
add the ability to directly compare mesh and forcing objects with `==`

### CODE REVIEW

File: adcircpy/mesh/base.py [hunk 1]
```diff
@@ -116,6 +116,12 @@ def to_dict(self):
         nodes = {nid: (coo, val) for nid, coo, val in zip(self._id, self._coords, self.values)}
         return nodes
 
+    def __eq__(self, other: 'Nodes') -> bool:
+        try:
+            return np.all(self.coords == other.coords) and np.all(self.values == other.values)
```
Comment (krober10nd) [inline adcircpy/mesh/base.py:121]: should this be to a certain level of precision?

File: tests/test_adcirc_mesh.py [hunk 2]
```diff
@@ -168,3 +168,17 @@ def test_make_plot_flat_domain(nodes, elements, boundaries, mocker):
 
     mocker.patch('matplotlib.pyplot.show')
     mesh.make_plot()
+
+
+def test_mesh_equality(nodes, elements, boundaries, wet_nodes, wet_elements):
+    mesh_1 = AdcircMesh(nodes, elements, boundaries)
+    mesh_2 = AdcircMesh(nodes, elements, boundaries)
+    mesh_3 = mesh_1
```
Comment (krober10nd) [inline tests/test_adcirc_mesh.py:177]: shouldn't this be a `copy()` or a `deepcopy()`?

File: tests/test_adcirc_mesh.py [hunk 3]
```diff
@@ -168,3 +168,17 @@ def test_make_plot_flat_domain(nodes, elements, boundaries, mocker):
...
+    mesh_4 = AdcircMesh(wet_nodes, wet_elements)
+
+    assert mesh_1 is not mesh_2
+    assert mesh_1 is mesh_3
```
Comment (krober10nd) [inline tests/test_adcirc_mesh.py:183]: what's the difference between `is` and `==`?

File: adcircpy/mesh/base.py (hunk 1 above)
Comment (jreniel) [reply adcircpy/mesh/base.py:121]: I'd say that even the slightest precision differences should make the mesh categorically different.

**Comment by jreniel** [comment]
Hi @zacharyburnettNOAA, I have a general question. What will be the result if the mesh graphs are identical but the model boundary definitions are different? (Model boundaries can be defined in many different ways for any given topology.)

File: tests/test_adcirc_mesh.py (hunk 3 above)
Comment (ghost) [reply tests/test_adcirc_mesh.py:183]: `is` checks if the two references refer to the same object (i.e. the literal same block in memory), whereas `==` will call `.__eq__()`

**Comment by ghost** [comment]
> Hi @zacharyburnettNOAA, I have a general question. What will be the result if the mesh graphs are identical but the model boundary definitions are different? (Model boundaries can be defined in many different ways for any given topology.)

good point, I forgot about the boundaries. Which object are they stored in again?

**Comment by jreniel** [comment]
> > Hi @zacharyburnettNOAA, I have a general question. What will be the result if the mesh graphs are identical but the model boundary definitions are different? (Model boundaries can be defined in many different ways for any given topology.)
> 
> good point, I forgot about the boundaries. Which object are they stored in again?

The base object only contains graph data, the fort.14 extends that to include boundaries, and AdcircMesh extends it further to include nodal attributes.

**Review by krober10nd — CHANGES_REQUESTED** [review]
My only general comment or thinking point would what Jaime said, how this can generalize to the other classes and aspects of the mesh like the boundaries and nodal attributes file. 

It would also be nice to make this a command line interface at some point. Before I burn a couple thousand dollars on a cloud machine, I'd probably like the code to assert out if I'm not simulating with the correct model.

**Comment by krober10nd** [comment]
Another test could be read in a complex for.13, write it out and read it and check it's the same. This bit me before. In fact, probably should do this for all I/O operations.

**Comment by jreniel** [comment]
It's not guaranteed to be the same, but bugs forgiving, they are guaranteed to be equivalent. If you expand both, they arrays themselves must be identical. If you want the NodalAtributes class to guarantee the exact same output as the input, you can modify the existing implementation to do it.

**Comment by krober10nd** [comment]
> It's not guaranteed to be the same, but bugs forgiving, they are guaranteed to be equivalent. If you expand both, they arrays themselves must be identical. If you want the NodalAtributes class to guarantee the exact same output as the input, you can modify the existing implementation to do it.

Yea I just mean to check that the IO implementation is always working. Work with a fort.13 that should always remain the same (i.e., has your median compression technique already applied).

**Comment by jreniel** [comment]
It is being applied, I only fixed a bug in the indexing. But if you really want your output to be identical you are welcome to provide the alternate implementation for it.

**Comment by krober10nd** [comment]
> It is being applied, I only fixed a bug in the indexing. But if you really want your output to be identical you are welcome to provide the alternate implementation for it.

Sorry, but you're misunderstanding what I'm saying. I'm trying to say that as a new unit test for `adcircpy` that covers both the file IO and, at the same time, tests the equality operator, to read in and write out files and then compare their equivalence with this new method.

**Review by jreniel — APPROVED** [review]
Looks good to me.

File: tests/test_adcirc_mesh.py (hunk 2 above)
Comment (ghost) [reply tests/test_adcirc_mesh.py:177]: added

**Review by krober10nd — APPROVED** [review]

**Comment by krober10nd** [comment]
Thanks!

**Comment by ghost** [comment]
one thing I did notice was that the `.to_dict()` method of `fort14` returns values as negative:
https://github.com/jreniel/adcircpy/blob/99fb3c1a508b46c7f32e85c67beee6f16510f914/adcircpy/mesh/fort14.py#L234-L245

To pass the copy tests, I changed `Nodes.__eq__()` to check by absolute values (99fb3c1):
https://github.com/jreniel/adcircpy/blob/99fb3c1a508b46c7f32e85c67beee6f16510f914/adcircpy/mesh/base.py#L119-L125

**Comment by jreniel** [comment]
> one thing I did notice was that the `.to_dict()` method of `fort14` returns values as negative:
> 
> https://github.com/jreniel/adcircpy/blob/99fb3c1a508b46c7f32e85c67beee6f16510f914/adcircpy/mesh/fort14.py#L234-L245
> 
> To pass the copy tests, I changed `Nodes.__eq__()` to check by absolute values ([99fb3c1](https://github.com/jreniel/adcircpy/commit/99fb3c1a508b46c7f32e85c67beee6f16510f914)):
> 
> https://github.com/jreniel/adcircpy/blob/99fb3c1a508b46c7f32e85c67beee6f16510f914/adcircpy/mesh/base.py#L119-L125

This is on purpose,  the fort.14 ascii files is positive down while the python arrays are positive up. You must be careful with that. Checking absolute values is not really what you want.

**Comment by ghost** [comment]
> This is on purpose, the fort.14 ascii files is positive down while the python arrays are positive up. You must be careful with that. Checking absolute values is not really what you want.

ah, ok, so perhaps it should flip the sign upon reading into memory instead of in `.to_dict()`?

**Comment by jreniel** [comment]
> > This is on purpose, the fort.14 ascii files is positive down while the python arrays are positive up. You must be careful with that. Checking absolute values is not really what you want.
> 
> ah, ok, so perhaps it should flip the sign upon reading into memory instead of in `.to_dict()`?

Yes, it should flip when reading a fort.14, but not when copying. Also be mindful that _only a fort.14_  is flipped. If you get an arbitrary file that is not a fort.14, but is has a fort.14 format (like for example, writing the fort.13 attributes as a mesh) those should not be  flipped. _Only the fort.14_ gets flipped upon read, then back upon write.

**Comment by jreniel** [comment]
> > > This is on purpose, the fort.14 ascii files is positive down while the python arrays are positive up. You must be careful with that. Checking absolute values is not really what you want.
> > 
> > 
> > ah, ok, so perhaps it should flip the sign upon reading into memory instead of in `.to_dict()`?
> 
> Yes, it should flip when reading a fort.14, but not when copying. Also be mindful that _only a fort.14_ is flipped. If you get an arbitrary file that is not a fort.14, but is has a fort.14 format (like for example, writing the fort.13 attributes as a mesh) those should not be flipped. _Only the fort.14_ gets flipped upon read, then back upon write.

For further clarification, the Fort14 class should only be used to read fort.14 but if you have such arbitrary file you should use the base class to read it instead (cause it's technically not a fort.14).

On this note it would be useful if we could return Grd objects for each nodal attribute so we can plot them or write them as grd files, which it's not necessary for the ADCIRC workflow, but it's useful for debugging and checking. This is something that might interest @krober10nd

**Comment by krober10nd** [comment]
```
On this note it would be useful if we could return Grd objects for each nodal attribute so we can plot them or write them as grd files, which it's not necessary for the ADCIRC workflow, but it's useful for debugging and checking. This is something that might interest @krober10nd
```

Yes, this could be useful although how would it work for the vectorial nodal attributes? Bring one attribute per direction?

**Comment by jreniel** [comment]
> ```
> On this note it would be useful if we could return Grd objects for each nodal attribute so we can plot them or write them as grd files, which it's not necessary for the ADCIRC workflow, but it's useful for debugging and checking. This is something that might interest @krober10nd
> ```
> 
> Yes, this could be useful although how would it work for the vectorial nodal attributes? Bring one attribute per direction?

We'll need to return either 2 grids or do some special treatment for vectors... Then there's the vectorial timeseries. The outputs API has some boilerplate for handling these, but in general, there could be lot's of improvements for the NodalAttributes API related to this.

**Comment by jreniel** [comment]
> how would it work for the vectorial nodal attributes?

You hit the nail here about the limitations of "grd" format, and why it is generally not supported for most mesh parsers. UGrid format, for example, handles this. And when it comes to these advanced files and needs, we need to adapt. We don't need to change the fort.14 format, but we'll need to work with other format types in order to really be able to handle the data.

This is related to the observations I make here: https://github.com/jreniel/adcircpy/blob/main/adcircpy/mesh/parsers/grd.py#L19-L25

**Pull request merged**